README.md
# Rack::LtsvLogger

A rack middleware to output access log in ltsv format, like [rack/commonlogger](https://github.com/rack/rack/blob/master/lib/rack/commonlogger.rb) (which output access log in apache common format).

## Why Rack Middleware

cf. https://speakerdeck.com/mirakui/high-performance-rails-long-edition

<img src="doc/x_runtime.png" alt="x_runtime" width="50%" height="50%"/>

The Completed Time, which the default logger of rails shows,
does not include the routing time, and the elapsed time on rack middleware layers.
To measure the processing time accurately,
it is necessary to insert a rack middleware at the head of rack middleware stacks.

## Installation

Add this line to your application's Gemfile:

    gem 'rack-ltsv_logger'

And then execute:

    $ bundle

## How to Use

Insert Rack::LtsvLogger on the head of rack middlewares. 

### Rails

```ruby
# config/application.rb
class Application < Rails::Application
  config.middleware.insert_after(0, Rack::LtsvLogger, $stdout)
end
```

Middleware check.

```
bundle exec rake middleware
```

### Sinatra

```ruby
# config.ru
require 'rack/ltsv_logger'

use Rack::LtsvLogger, $stdout
run App
```

## Format

Sample (line feeded, but actually tab seperated):

```
time:2014-07-02T21:52:31+09:00
pid:15189
host:127.0.0.1
forwardedfor:127.0.0.2
user:user
method:GET
uri:/get
query:?foo
protocol:HTTP/1.1
status:200
size:-
reqtime:0.000000
```

### Default Fields

| key          | value                                                                          |
|--------------|--------------------------------------------------------------------------------|
| time         | The datetime in ISO-8601 format                                                |
| pid          | Procedd ID                                                                     |
| host         | ENV['REMOTE_ADDR']                                                             |
| forwardedfor | ENV['X_FORWARDED_FOR']                                                         |
| user         | ENV['REMOTE_USER']                                                             |
| method       | ENV['REQUEST_METHOD']                                                          |
| uri          | ENV['PATH_INFO']                                                               |
| query        | ENV['QUERY_STRING']                                                            |
| protocol     | ENV['HTTP_VERSION']                                                            |
| status       | Response Status Code                                                           |
| size         | Response Content-Length                                                        |
| reqtime      | The request time in secods. milli seconds are written after the decimal point. |


## Arguments

Arguments:

* io
  * An IO object, or something which has `#write` method
* ~~appends (deprecated)~~
  * ~~A Hash object to append custom fields. See [#custom-fields](#custom-fields)~~

Keyword Arguments:

* params\_proc
  * A Proc object to generate fields. See [#custom-fields](#custom-fields)

### Log Rotation

If you utilize log rotation functionality of Ruby's standard logger, you may write as

```ruby
class MyLogger < ::Logger
  def write(msg)
    @logdev.write msg
  end
end

logger = MyLogger.new('/path/to/access.log', 3, 1048576)
config.middleware.insert_after(0, Rack::LtsvLogger, logger)
```

### Custom Fields

Pass a proc object to customize fields as:

```ruby
params_proc = Proc.new do |env, status, headers, body, began_at|
  params = Rack::LtsvLogger::DEFAULT_PARAMS_PROC.call(env, status, headers, body, began_at)
  params.delete(:protocol)
  params.merge!(
    vhost: env['HTTP_HOST'] || "-",
    ua: env['HTTP_USER_AGENT'] || "-",
    referer: env['HTTP_REFERER'] || "-",
  )
end
config.middleware.insert_after(0, Rack::LtsvLogger, $stdout, params_proc: params_proc)
```

or

```ruby
params_proc = Proc.new do |env, status, headers, body, began_at|
  now = Time.now
  reqtime = now.instance_eval { to_i + (usec/1000000.0) } - began_at
  {
    time: now.iso8601,
    pid: Process.pid,
    host: env["REMOTE_ADDR"] || "-",
    forwardedfor: env['HTTP_X_FORWARDED_FOR'] || "-",
    user: env["REMOTE_USER"] || "-",
    method: env["REQUEST_METHOD"],
    uri: env["PATH_INFO"],
    query: env["QUERY_STRING"].empty? ? "" : "?"+env["QUERY_STRING"],
    status: ::Rack::LtsvLogger.extract_status(status),
    size: ::Rack::LtsvLogger.extract_content_length(headers),
    reqtime: "%0.6f" % reqtime,
    vhost: env['HTTP_HOST'] || "-",
    ua: env['HTTP_USER_AGENT'] || "-",
    referer: env['HTTP_REFERER'] || "-",
  }
end
config.middleware.insert_after(0, Rack::LtsvLogger, $stdout, params_proc: params_proc)
```

I recommend to follow http://ltsv.org/ to add new key names.

**Deprecated**

```ruby
appends = {
  vhost: Proc.new {|env| env['HTTP_HOST'] || "-" },
  ua: Proc.new {|env| env['HTTP_USER_AGENT'] || "-" },
  referer: Proc.new {|env| env['HTTP_REFERER'] || "-" },
}
config.middleware.insert_after(0, Rack::LtsvLogger, $stdout, appends)
```

## ChangeLog

See [CHANGELOG.md](CHANGELOG.md) for details.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new [Pull Request](../../pull/new/master)

## Copyright

See [LICENSE.txt](LICENSE.txt) for details.


CHANGELOG.md
## 0.1.1  (2019/02/08)

Fixes:

* Status field includes exactly 3 chracters (thanks to SpringMT)

## 0.1.0  (2017/11/26)

Enhancements:

* Add params\_proc keyword arguments to support more flexible ways to customize fields.

Changes:

* `appends` argument is deprecated.

## 0.0.3  (2016/04/03)

Fixes:

* Fix pid to be varing

## 0.0.2  (2016/02/19)

Enhancements:

* Ensure to log even if `@app.call` raises an error (thanks to @sugilog)

## 0.0.1  (2014/07/02)

First release


spec/spec_helper.rb
require 'bundler'
Bundler.setup(:default, :test)
Bundler.require(:default, :test)

#require 'simplecov'
#require 'simplecov-rcov'
#SimpleCov.formatter = SimpleCov::Formatter::RcovFormatter
#SimpleCov.start

$TESTING=true
$:.unshift File.join(File.dirname(__FILE__), '..', 'lib/rack/')
require 'timecop'
require 'rack'
require 'rack-ltsvlogger'


spec/ltsv_logger_spec.rb
require_relative 'spec_helper'
require 'timecop'

describe Rack::LtsvLogger do
  def app
    lambda { |env|
      [200, {'Content-Type' => 'text/plain'}, ["Hello, World!"]]
    }
  end

  def error_app
    lambda { |env|
      raise ArgumentError, "error test"
    }
  end

  def parse_ltsv(ltsv)
    Hash[*(ltsv.chomp.split("\t").map {|e| e.split(":", 2) }.flatten)]
  end

  let(:now)      { Time.now }
  let(:time)     { now.iso8601 }
  let(:pid)      { Process.pid }
  let(:host)     { '127.0.0.1' }
  let(:forwardedfor) { '127.0.0.2' }
  let(:user)     { 'user' }
  let(:protocol) { 'HTTP/1.1' }
  let(:env) do
    {
      'REMOTE_ADDR' => host,
      'HTTP_X_FORWARDED_FOR' => forwardedfor,
      'REMOTE_USER' => user,
      'HTTP_VERSION' => protocol,
    }
  end

  before do
    Timecop.freeze now
  end

  after do
    Timecop.return
  end

  context 'default ltsv' do
    subject do
      @output = StringIO.new
      Rack::Lint.new( Rack::LtsvLogger.new(app, @output) )
    end

    let(:method)   { 'GET' }
    let(:uri)      { '/get' }
    let(:query)    { '?foo' }

    it do
      Rack::MockRequest.new(subject).get("#{uri}#{query}", env)
      expect(@output.string).to eq \
        "time:#{time}\tpid:#{pid}\thost:#{host}\tforwardedfor:#{forwardedfor}\tuser:#{user}\t" +
        "method:#{method}\turi:#{uri}\tquery:#{query}\tprotocol:#{protocol}\tstatus:200\tsize:-\treqtime:0.000000\n"
    end
  end

  context 'custom fields' do
    let(:params_proc) do
      Proc.new do |env, status, headers, body, began_at|
        params = Rack::LtsvLogger::DEFAULT_PARAMS_PROC.call(env, status, headers, body, began_at)
        params.delete(:protocol)
        params.merge!(
          new_protocol: env['HTTP_VERSION'],
        )
      end
    end

    subject do
      @output = StringIO.new
      Rack::Lint.new( Rack::LtsvLogger.new(app, @output, params_proc: params_proc) )
    end

    it 'GET /get' do
      Rack::MockRequest.new(subject).get('/get', env)
      params = parse_ltsv(@output.string)
      expect(params['protocol']).to be_nil
      expect(params['new_protocol']).to eq(protocol)
    end
  end

  context 'appends (deprecated)' do
    let(:appends) do
      { x_runtime: Proc.new {|env| '1.234' } }
    end

    subject do
      @output = StringIO.new
      Rack::Lint.new( Rack::LtsvLogger.new(app, @output, appends) )
    end

    it 'GET /get' do
      Rack::MockRequest.new(subject).get('/get', env)
      params = parse_ltsv(@output.string)
      expect(params['x_runtime']).to eq('1.234')
    end
  end 

  context 'when app error occured' do
    let(:appends) do
      { x_runtime: Proc.new {|env| '1.234' } }
    end

    subject do
      @output = StringIO.new
      Rack::Lint.new( Rack::LtsvLogger.new(error_app, @output, appends) )
    end

    it 'GET /get' do
      expect {
        Rack::MockRequest.new(subject).get('/get', env)
      }.to raise_error ArgumentError

      params = parse_ltsv(@output.string)
      expect(params).not_to be_empty
      expect(params['x_runtime']).to eq('1.234')
      expect(params['status']).to eq('500')
    end
  end

  context 'when app returns invalid http status code' do
    subject do
      @output = StringIO.new
      invalid_status_code_app = lambda do |env|
        [2000, {'Content-Type' => 'text/plain'}, ["Hello, World!"]]
      end
      Rack::Lint.new( Rack::LtsvLogger.new(invalid_status_code_app, @output) )
    end

    it 'GET /get' do
      Rack::MockRequest.new(subject).get('/get', env)
      params = parse_ltsv(@output.string)
      expect(params['status']).to eq '200'
    end
  end
end


spec/rack-ltsvlogger_spec.rb
require_relative 'spec_helper'
require 'fileutils'

describe Rack::LtsvLogger do
  app = lambda { |env|
    [200, {'Content-Type' => 'text/plain'}, ["Hello, World!"]]
  }

  let(:log_dir)  { "#{File.dirname(__FILE__)}/log" }
  let(:now)      { Time.now }
  let(:time)     { now.iso8601 }
  let(:pid)      { Process.pid }
  let(:host)     { '127.0.0.1' }
  let(:vhost)    { '127.0.0.1:80' }
  let(:forwardedfor) { '127.0.0.2' }
  let(:ua)       { 'mock' }
  let(:referer)  { 'http://example.com' }
  let(:env) do
    {
      'REMOTE_ADDR' => host,
      'HTTP_HOST' => vhost,
      'HTTP_X_FORWARDED_FOR' => forwardedfor,
      'HTTP_USER_AGENT' => ua,
      'HTTP_REFERER' => referer,
    }
  end

  def parse_ltsv(ltsv)
    Hash[*(ltsv.chomp.split("\t").map {|e| e.split(":", 2) }.flatten)]
  end

  def ltsv(hash)
    hash.map {|k, v| "#{k}:#{v}" }.join("\t") + "\n"
  end

  before do
    Dir.mkdir log_dir
    Timecop.freeze now
  end

  after do
    FileUtils.rm_rf log_dir
    Timecop.return
  end

  context 'conform to Rack::Lint' do
    subject do
      @test_io = StringIO.new
      Rack::Lint.new( Rack::LtsvLogger.new(app, @test_io) )
    end

    context 'GET /get?foo' do
      let(:method) { 'GET' }
      let(:uri)    { '/get' }
      it do
        Rack::MockRequest.new(subject).get('/get?foo', env)
        expect(@test_io.string).to eq \
          "time:#{time}\tpid:#{pid}\thost:#{host}\tvhost:#{vhost}\tforwardedfor:#{forwardedfor}\t" +
          "size:-\tstatus:200\tmethod:#{method}\turi:#{uri}\tua:#{ua}\treferer:#{referer}\treqtime:0.0\n"
      end
    end

    context 'POST /post' do
      let(:method) { 'POST' }
      let(:uri)    { '/post' }
      it do
        Rack::MockRequest.new(subject).post('/post', env)
        expect(@test_io.string).to eq \
          "time:#{time}\tpid:#{pid}\thost:#{host}\tvhost:#{vhost}\tforwardedfor:#{forwardedfor}\t" +
          "size:-\tstatus:200\tmethod:#{method}\turi:#{uri}\tua:#{ua}\treferer:#{referer}\treqtime:0.0\n"
      end
    end
  end

  context 'append fields' do
    let(:appends) do
      { x_runtime: Proc.new {|env| '1.234' } }
    end

    subject do
      @test_io = StringIO.new
      Rack::Lint.new( Rack::LtsvLogger.new(app, @test_io, appends) )
    end

    it 'GET /get' do
      Rack::MockRequest.new(subject).get('/get', env)
      params = parse_ltsv(@test_io.string)
      expect(params['x_runtime']).to eq('1.234')
    end
  end
end


lib/rack/ltsvlogger.rb
require 'time'

module Rack
  class LtsvLogger
    PID = Process.pid

    def initialize(app, logger=nil, appends = {})
      @app = app
      @logger = logger || $stdout
      @appends = appends
    end

    def call(env)
      start_time = Time.now.instance_eval { to_i + (usec/1000000.0) }

      status, headers, body = @app.call(env)

      now = Time.now
      time = now.iso8601
      request_time = now.instance_eval { to_i + (usec/1000000.0) } - start_time

      params = {
        time: time,
        pid: PID,
        host: env['REMOTE_ADDR'] || "-",
        vhost: env['HTTP_HOST'] || "-",
        forwardedfor: env['HTTP_X_FORWARDED_FOR'] || "-",
        size: extract_content_length(headers),
        status: status,
        method: env['REQUEST_METHOD'] || "-",
        uri: env['PATH_INFO'] || "-",
        ua: env['HTTP_USER_AGENT'] || "-",
        referer: env['HTTP_REFERER'] || "-",
        reqtime: request_time,
      }
      @appends.each do |key, proc|
        params[key] = proc.call(env)
      end
      @logger.write ltsv(params)

      [status, headers, body]
    end

    private

    def ltsv(hash)
      hash.map {|k, v| "#{k}:#{v}" }.join("\t") + "\n"
    end

    def extract_content_length(headers)
      value = headers['Content-Length'] or return '-'
      value.to_s == '0' ? '-' : value
    end
  end
end


lib/rack/ltsv_logger.rb
require 'time'

module Rack
  class LtsvLogger
    DEFAULT_PARAMS_PROC = Proc.new do |env, status, headers, body, began_at|
      now = Time.now
      reqtime = now.instance_eval { to_i + (usec/1000000.0) } - began_at
      {
        time: now.iso8601,
        pid: Process.pid,
        host: env["REMOTE_ADDR"] || "-",
        forwardedfor: env['HTTP_X_FORWARDED_FOR'] || "-",
        user: env["REMOTE_USER"] || "-",
        method: env["REQUEST_METHOD"],
        uri: env["PATH_INFO"],
        query: env["QUERY_STRING"].empty? ? "" : "?"+env["QUERY_STRING"],
        protocol: env["HTTP_VERSION"],
        status: ::Rack::LtsvLogger.extract_status(status),
        size: ::Rack::LtsvLogger.extract_content_length(headers),
        reqtime: "%0.6f" % reqtime,
      }
    end

    def initialize(app, io = nil, **kwargs)
      @app = app
      @io = io || $stdout
      @params_proc = kwargs[:params_proc] || DEFAULT_PARAMS_PROC
      @appends = kwargs.tap {|h| h.delete(:params_proc) } # old version compatibility
    end

    def call(env)
      began_at = Time.now.instance_eval { to_i + (usec/1000000.0) }
      status, headers, body = @app.call(env)
    ensure
      params = @params_proc.call(env, status, headers, body, began_at)
      @appends.each {|key, proc| params[key] = proc.call(env) } # old version compatibility
      @io.write ltsv(params)
    end

    def self.extract_content_length(headers)
      value = headers && headers['Content-Length'] or return '-'
      value.to_s == '0' ? '-' : value
    end

    def self.extract_status(status)
      status.nil? ? "500" : status.to_s[0..2]
    end

    private

    def ltsv(hash)
      hash.map {|k, v| "#{k}:#{v}" }.join("\t") + "\n"
    end
  end
end
